### PR TITLE
 Change EventJournal Projection to Function

### PR DESCRIPTION
**Pull request by viliam-durina**
Also add default implementations for use by Jet

### CODE REVIEW

File: hazelcast/src/main/java/com/hazelcast/map/impl/journal/MapEventJournalFunctions.java [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+/**
+ * Utility functions used from Jet. They are here so that they can be also used
+ * when Jet accesses remote imdg (non-jet) cluster.
+ */
+public final class MapEventJournalFunctions {
+
+    private MapEventJournalFunctions() { }
+
+    public static <K, V> Predicate<EventJournalMapEvent<K, V>> mapPutEvents() {
+        return new MapPutEventsPredicate<>();
```
Comment (mmedenjak) [inline hazelcast/src/main/java/com/hazelcast/map/impl/journal/MapEventJournalFunctions.java]: Although we require JDK8, we must remain JDK6 compatible so you must specify the types here.

File: hazelcast/src/main/java/com/hazelcast/map/impl/journal/MapEventJournalFunctions.java [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+    }
+
+    public static <K, V> Function<EventJournalMapEvent<K, V>, Entry<K, V>> mapEventToEntry() {
+        return new MapEventToEntryProjection<>();
```
Comment (mmedenjak) [inline hazelcast/src/main/java/com/hazelcast/map/impl/journal/MapEventJournalFunctions.java]: Same as above.

File: hazelcast/src/main/java/com/hazelcast/internal/journal/EventJournalDataSerializerHook.java [hunk 3]
```diff
@@ -52,6 +53,8 @@ public IdentifiedDataSerializable create(int typeId) {
                 switch (typeId) {
                     case EVENT_JOURNAL_INITIAL_SUBSCRIBER_STATE:
                         return new EventJournalInitialSubscriberState();
+                    case DESERIALIZING_ENTRY:
+                        return new DeserializingEntry<>();
```
Comment (mmedenjak) [inline hazelcast/src/main/java/com/hazelcast/internal/journal/EventJournalDataSerializerHook.java]: Remove the diamond operator.

File: hazelcast/src/main/java/com/hazelcast/cache/impl/journal/CacheEventJournalFunctions.java [hunk 4]
```diff
@@ -0,0 +1,82 @@
...
+/**
+ * Utility functions used from Jet. They are here so that they can be also used
+ * when Jet accesses remote imdg (non-jet) cluster.
+ */
+public final class CacheEventJournalFunctions {
+
+    private CacheEventJournalFunctions() { }
+
+    public static <K, V> Predicate<EventJournalCacheEvent<K, V>> cachePutEvents() {
+        return new CachePutEventsPredicate<>();
```
Comment (mmedenjak) [inline hazelcast/src/main/java/com/hazelcast/cache/impl/journal/CacheEventJournalFunctions.java]: Missing types

File: hazelcast/src/main/java/com/hazelcast/cache/impl/journal/CacheEventJournalFunctions.java [hunk 5]
```diff
@@ -0,0 +1,82 @@
...
+    }
+
+    public static <K, V> Function<EventJournalCacheEvent<K, V>, Entry<K, V>> cacheEventToEntry() {
+        return new CacheEventToEntryProjection<>();
```
Comment (mmedenjak) [inline hazelcast/src/main/java/com/hazelcast/cache/impl/journal/CacheEventJournalFunctions.java]: Missing types

File: hazelcast/src/test/java/com/hazelcast/journal/AbstractEventJournalBasicTest.java [hunk 6]
```diff
@@ -68,13 +67,12 @@
...
 public abstract class AbstractEventJournalBasicTest<EJ_TYPE> extends HazelcastTestSupport {
     private static final Random RANDOM = new Random();
-    private static final String LIMITED_DS = "limited";
 
     protected HazelcastInstance[] instances;
 
     private int partitionId;
     private TruePredicate<EJ_TYPE> TRUE_PREDICATE = new TruePredicate<EJ_TYPE>();
-    private Projection<EJ_TYPE, EJ_TYPE> IDENTITY_PROJECTION = Projections.identity();
+    private Function<EJ_TYPE, EJ_TYPE> IDENTITY_FUNCTION = new IdentityFunction<>();
```
Comment (mmedenjak) [inline hazelcast/src/test/java/com/hazelcast/journal/AbstractEventJournalBasicTest.java]: Missing type

File: hazelcast/src/test/java/com/hazelcast/journal/AbstractEventJournalExpiringTest.java [hunk 7]
```diff
@@ -46,7 +45,7 @@
 
     private int partitionId;
     private TruePredicate<EJ_TYPE> TRUE_PREDICATE = new TruePredicate<EJ_TYPE>();
-    private Projection<EJ_TYPE, EJ_TYPE> IDENTITY_PROJECTION = Projections.identity();
+    private Function<EJ_TYPE, EJ_TYPE> IDENTITY_FUNCTION = new IdentityFunction<>();
```
Comment (mmedenjak) [inline hazelcast/src/test/java/com/hazelcast/journal/AbstractEventJournalExpiringTest.java]: Missing type

File: hazelcast/src/test/java/com/hazelcast/journal/AbstractEventJournalBounceTest.java [hunk 8]
```diff
@@ -117,7 +116,7 @@ public void run() {
                 final ReadResultSet<T> partitionEvents = reader.readFromEventJournal(
                         state.getOldestSequence(), 1,
                         (int) (state.getNewestSequence() - state.getOldestSequence() + 1), i,
-                        new TruePredicate<T>(), Projections.<T>identity()).get();
+                        new TruePredicate<T>(), new IdentityFunction<>()).get();
```
Comment (mmedenjak) [inline hazelcast/src/test/java/com/hazelcast/journal/AbstractEventJournalBounceTest.java]: Missing type

**Comment by ahmetmircik** [comment]
Seems change to Function broke EE compatibility tests: http://jenkins.hazelcast.com/job/Hazelcast-EE-pr-builder/332/testReport/junit/com.hazelcast.test.compatibility/SerializedObjectsCompatibilityTest/testObjectsAreDeserializedInCurrentVersion_whenEESerializationService_samplesVersion__3_10_testDeserializerVersion_3_10_/

**Comment by vbekiaris** [comment]
see #14672

**Pull request merged**